---
title:  "Saureus methylation"
author: "Natalia Zajac"
output:
  html_document: 
    highlight: pygments
    theme: sand
    code_folding: hide
    toc: yes
    toc_depth: 4
editor_options: 
  chunk_output_type: inline
---

Started on `r format(Sys.time(), "%Y-%m-%d %H:%M:%S")`

```{r setup, echo=FALSE, warning=FALSE, include=FALSE}
library(readr)
library(tidyverse)
library(rtracklayer)
library(dplyr)
library(reshape2)
library(tidyverse)
library(cowplot)
library(patchwork)
library(data.table)
library(ShortRead)
library(ezRun)

sample_metadata = read.delim("/srv/GT/analysis/zajacn/course_RCIG/methylation_sample_metadata.txt")
sample_metadata$condition = sapply(str_split(sample_metadata$Name, "_"), .subset, 2)
geneAnno = readGFF("/srv/GT/analysis/zajacn/course_RCIG/SA6850_set_forIGV/SA_6850_GCA_000462955.1_ASM46295v1_genomic.gff")
geneAnno = geneAnno[geneAnno$type == "gene",]
fasta6850 = readDNAStringSet("SA6850_set_forIGV/SA_6850_ncbi_sequence.fasta")
bases = data.frame(count = sapply(list("C", "A", "G", "T"), function(base) vcountPattern(base,fasta6850))) %>% rownames_to_column("Sequence")

# Read GFF file 
methylation = NULL
for (i in sample_metadata$Name){
  smp = sample_metadata[sample_metadata$Name == i,]
  methylation_data <- read_delim(smp$path_to_met_result, delim = "\t", comment = "#", 
                               col_names = c("seqname", "source", "feature", "start", 
                                             "end", "score", "strand", "phase", "attributes"))
  
  #x = read.csv(str_replace(smp$path_to_met_result, "motifs.gff", "basemods.csv"))

  # Extract key attributes (IPD ratio, modification type, etc.)
  methylation_data <- methylation_data %>%
  mutate(
    motif = str_extract(attributes, "context=[^;]+") %>% str_remove("context="),
    base_after = substr(motif, 21,22),
    base_before = substr(motif, 20,21),
    coverage = str_extract(attributes, "coverage=[^;]+") %>% str_remove("coverage=") %>% as.numeric(),
    IPDRatio = str_extract(attributes, "IPDRatio=[^;]+") %>% str_remove("IPDRatio=") %>% as.numeric(),
    score = as.numeric(score),
    identificationQv = str_extract(attributes, "identificationQv=[^;]+") %>% str_remove("identificationQv=") %>% as.numeric()
  )

  filtered_data <- methylation_data %>%
    filter(score > 30) %>%
    filter(coverage > 200) 
  
  #Look for overlap with genes
  minus_strand = filtered_data[filtered_data$strand == "-",]
  plus_strand = filtered_data[filtered_data$strand == "+",]
  
  #Minus
  genes <- as.data.table(unique(geneAnno[, c("seqid", "start", "end", "strand")]))
  genes = genes[genes$strand == "-",c("seqid", "start", "end")]
  genes$upstream = genes$start-200
  genes = genes %>% mutate(upstream = if_else(upstream < 0, 0, upstream))
  genes$downstream = genes$end+200
  genes = genes %>% mutate(downstream = if_else(downstream > length(fasta6850$`CP006706.1 Staphylococcus aureus subsp. aureus 6850, complete genome`), length(fasta6850$`CP006706.1 Staphylococcus aureus subsp. aureus 6850, complete genome`), downstream))
  setDT(genes, key = c("seqid", "upstream", "downstream"))
  
  metbases <- as.data.table(unique(minus_strand[, c("seqname", "start", "end")]))
  colnames(metbases) = c("seqid", "snp", "pos2")
  setDT(metbases, key = c("seqid", "snp", "pos2"))
  
  obj <- foverlaps(metbases, genes, by.x = c("seqid", "snp", "pos2"),
                 by.y = c("seqid", "upstream", "downstream"),
                 type = "within", nomatch = 0)
  obj <- merge(minus_strand, obj,
             by.y = c("seqid", "snp", "pos2"),
             by.x = c("seqname", "start", "end"),
             all.x = TRUE)
  obj <- merge(obj,
             geneAnno[geneAnno$strand == "-", c("seqid", "start", "end", "ID" , "gene", "locus_tag")],
             by.y = c("seqid", "start", "end"),
             by.x = c("seqname", "start.y", "end.y"),
             all.x = TRUE)
  minus = data.frame(obj)
  colnames(minus) = c("seqname", "gene_start", "gene_end", "start", "end",colnames(minus)[6:22])
  
  #Plus
  genes <- as.data.table(unique(geneAnno[, c("seqid", "start", "end", "strand")]))
  genes = genes[genes$strand == "+",c("seqid", "start", "end")]
  genes$upstream = genes$start-200
  genes = genes %>% mutate(upstream = if_else(upstream < 0, 0, upstream))
  genes$downstream = genes$end+200
  genes = genes %>% mutate(downstream = if_else(downstream > length(fasta6850$`CP006706.1 Staphylococcus aureus subsp. aureus 6850, complete genome`), length(fasta6850$`CP006706.1 Staphylococcus aureus subsp. aureus 6850, complete genome`), downstream))
  setDT(genes, key = c("seqid", "upstream", "downstream"))
  
  metbases <- as.data.table(unique(plus_strand[, c("seqname", "start", "end")]))
  colnames(metbases) = c("seqid", "snp", "pos2")
  setDT(metbases, key = c("seqid", "snp", "pos2"))
  
  obj <- foverlaps(metbases, genes, by.x = c("seqid", "snp", "pos2"),
                 by.y = c("seqid", "upstream", "downstream"),
                 type = "within", nomatch = 0)
  obj <- merge(plus_strand, obj,
             by.y = c("seqid", "snp", "pos2"),
             by.x = c("seqname", "start", "end"),
             all.x = TRUE)
  obj <- merge(obj,
             geneAnno[geneAnno$strand == "+", c("seqid", "start", "end", "ID" , "gene", "locus_tag")],
             by.y = c("seqid", "start", "end"),
             by.x = c("seqname", "start.y", "end.y"),
             all.x = TRUE)
  plus = data.frame(obj)
  colnames(plus) = c("seqname", "gene_start", "gene_end", "start", "end",colnames(plus)[6:22])
  
  obj = rbind(minus,plus)
  print(length(unique(obj$start)) == length(unique(filtered_data$start)))
  
  obj = obj %>% mutate(category = case_when(start < gene_start & start >= upstream ~ "upstream", start > gene_end & start <= downstream ~ "downstream", start >= gene_start & start <= gene_end ~ "gene", is.na(upstream) == TRUE ~ "intergenic"))
  
  methylation[[i]] = obj
}


positions6mA = lapply(methylation, function(x) as.character(unique(x[x$feature == "m6A",]$start)))
positions4mC = lapply(methylation, function(x) as.character(unique(x[x$feature == "m4C",]$start)))

mtx = bind_rows(methylation, .id = "Name")

sequences <- readDNAStringSet("SA6850_set_forIGV/cds_from_genomic.fna")
names(sequences) = str_extract(names(sequences), "locus_tag=[^]]+")  %>% str_remove("locus_tag=") 
sequence_names <- names(sequences)
a_counts <- lapply(sequences, function(seq){ 
  matches <- gregexpr("A", seq)
  count <- sum(attr(matches[[1]], "match.length") > 0)
  length = nchar(seq)
  total = count #/length
  return(total)
})
a_counts_per_gene <- data.frame(Acounts = t(data.frame(a_counts))) %>% rownames_to_column()
c_counts <- lapply(sequences, function(seq){ 
  matches <- gregexpr("C", seq)
  count <- sum(attr(matches[[1]], "match.length") > 0)
  length = nchar(seq)
  total = count #/length
  return(total)
})
c_counts_per_gene <- data.frame(Ccounts = t(data.frame(c_counts))) %>% rownames_to_column()
ac_counts = merge(a_counts_per_gene,c_counts_per_gene, by = "rowname")
mtx = merge(mtx, ac_counts, by.x = "locus_tag", by.y = "rowname", all.x = TRUE)

names = setNames(if_else(grepl("A", sapply(str_split(unique(mtx$Name), "_"), .subset, 3)) == TRUE, sapply(str_split(unique(mtx$Name), "_"), .subset, 1), sapply(str_split(unique(mtx$Name), "_"), .subset, 1)), unique(mtx$Name))
names1 = setNames(rep("PASN", 9), unique(mtx$Name)[grepl("SN", unique(mtx$Name))])
names2 = setNames(rep("TSB", 9), unique(mtx$Name)[grepl("TSB", unique(mtx$Name))])
treatment = c(names1, names2)
metad = merge(data.frame(treatment) %>% rownames_to_column(), data.frame(names) %>% rownames_to_column(), by = "rowname")
metad$replicate = sapply(str_split(sapply(str_split(metad$rowname, "_"), .subset, 3), ""), .subset, 2) %>% as.numeric()
colnames(metad) = c("Name", "treatment", "group", "replicate")
mtx = merge(mtx, metad, by = "Name", all.x = TRUE)

#Limit it to methylation sites that are present in at least 66% of replicates

at_least_two_reps = mtx %>% group_by(treatment, group, feature, start, strand) %>% summarise(replicates = n_distinct(replicate)) %>% filter(replicates > 1)
at_least_two_reps$keep = "Keep"
mtx = merge(mtx, at_least_two_reps[,-6], by = c("treatment", "group", "start", "strand", "feature"), all.x = TRUE)
mtx = mtx[which(mtx$keep == "Keep"),]

```


Explanation of all the statistics given by ipdSummary:

motifs.gff
score - Phred-transformed pvalue that a kinetic deviation exists at this position
IPDratio - interpulse duration ratio: capped mean of normalized IPDs observed at this position/	normalized mean IPD predicted by the synthetic control model for this sequence context, > 1 - the sequencing polymerase slowed down, < 1 the sequencing polymerase sped up
context - 20bp before and 20bp after the modified base
identificationQv - the phred-transformed probability of an incorrect identification. Low indicates higher probability of an error.
coverage - count of valid IPDs at this position

basemods.csv
tpl - 1-based template position
tMean - capped mean of normalized IPDs observed at this position
modelPrediction - normalized mean IPD predicted by the synthetic control model for this sequence context

For whole genome amplification, the statistical test used is a case-control, two-sample t-test.

https://github.com/PacificBiosciences/kineticsTools/blob/master/doc/manual.rst

# {.tabset}

## QC {.tabset}

### Plotting IPDratio by score 

```{r, fig.height=5, fig.width=5, echo=FALSE, warning=FALSE}

for (i in sample_metadata$Name) {
  print(ggplot(methylation[[i]], aes(IPDRatio,score,colour = feature)) + geom_point())
}

```

### Plotting IPDratio by coverage


```{r, fig.height=5, fig.width=5, echo=FALSE, warning=FALSE}

for (i in sample_metadata$Name) {
  print(ggplot(methylation[[i]], aes(IPDRatio,coverage, colour = feature)) + geom_point())
}

```

### Plotting IPDratio by score categories

```{r, fig.height=5, fig.width=5, echo=FALSE, warning=FALSE}


for (i in sample_metadata$Name) {
  methylation[[i]]$IPDratioCats = ezCut(methylation[[i]]$IPDRatio, breaks = 1:8)
  methylation[[i]]$scoreCats = ezCut(methylation[[i]]$score, breaks = seq(0,2000,250))
  print(methylation[[i]] %>% group_by(scoreCats,IPDratioCats) %>% summarise(n_distinct(start)) %>% ggplot(aes(IPDratioCats, `n_distinct(start)`, fill = scoreCats)) + geom_col(position = "dodge") + scale_y_log10())
}


```

### Distributions

```{r, fig.height=6, fig.width=12, echo=FALSE, warning=FALSE}

p1 = ggplot(mtx[mtx$feature == "m6A",], aes(score)) + geom_density() 
p2 = ggplot(mtx[mtx$feature == "m6A",], aes(IPDRatio)) + geom_density() 

p1 + p2

```

## Analysis {.tabset}

### Number of 6mA and 4mC methylation sites per sample

```{r, fig.height=10, fig.width=12, echo=FALSE, warning=FALSE}

p1 = bind_rows(methylation, .id = "Name") %>% mutate(condition = sapply(str_split(Name, "_"), .subset, 2)) %>% group_by(Name, condition, feature, category) %>% summarise(count = n_distinct(start)) %>% ggplot(aes(Name, count, color = feature)) + geom_point() + facet_grid(category~condition, scales = "free") + theme(axis.text.x = element_text(angle = 45,hjust = 1, vjust = 1))

p2 = bind_rows(methylation, .id = "Name") %>% mutate(condition = sapply(str_split(Name, "_"), .subset, 2)) %>% group_by(Name, condition, feature, category) %>% summarise(count = n_distinct(start)) %>% ggplot(aes(condition, count, color = feature)) + geom_boxplot() + facet_grid(~category, scales = "free")

p1/p2
```

### Number of genes with 6mA or 4mC modifications

```{r, fig.height=7, fig.width=7, echo=FALSE, warning=FALSE}

p1 = bind_rows(methylation, .id = "Name") %>% mutate(condition = sapply(str_split(Name, "_"), .subset, 2)) %>% group_by(Name, condition, feature) %>% summarise(count = n_distinct(gene)) %>% ggplot(aes(condition, count, color = feature)) + geom_boxplot() 

p1
```


### Sharing of 6mA or 4mC modified sites

```{r, fig.height=10, fig.width=13, echo=FALSE, warning=FALSE}

ggVennDiagram::ggVennDiagram(positions6mA, nintersects = 100)

```


```{r, fig.height=10, fig.width=13, echo=FALSE, warning=FALSE}

ggVennDiagram::ggVennDiagram(positions4mC, nintersects = 100)

```

Methylated sites overlap between treatments filtered 

```{r, fig.height=10, fig.width=13, echo=FALSE, warning=FALSE}

ggVennDiagram::ggVennDiagram(list("PASN" = mtx[mtx$feature == "m6A"& mtx$treatment == "PASN",]$start, "TSB" = mtx[ mtx$feature == "m6A"& mtx$treatment == "TSB",]$start))

```

Heatmap of mtx


```{r, fig.height=10, fig.width=13, echo=FALSE, warning=FALSE}

ggVennDiagram::ggVennDiagram(list("PASN" = mtx[mtx$feature == "m6A"& mtx$treatment == "PASN",]$start, "TSB" = mtx[ mtx$feature == "m6A"& mtx$treatment == "TSB",]$start))

```


## Ideas:

-how the methylation is detected:
https://www.nature.com/articles/s41576-018-0081-3
https://www.pacb.com/wp-content/uploads/2015/09/WP_Detecting_DNA_Base_Modifications_Using_SMRT_Sequencing.pdf

-6mA density per gene: (but also 4mC)
https://link.springer.com/article/10.1186/s12864-024-10074-y#Sec10

-comparing ancestors (TSB vs PASN) + each clone
-comparing gene categories (methylation levels between categories - across all samples)

